File: app/src/main/java/com/pnuema/simplebible/ui/adapters/VersesAdapter.java
package com.pnuema.simplebible.ui.adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import com.pnuema.simplebible.data.IVerse;
import com.pnuema.simplebible.ui.viewholders.CopyrightViewHolder;
import com.pnuema.simplebible.ui.viewholders.VerseViewHolder;

import java.util.ArrayList;
import java.util.List;

public class VersesAdapter extends RecyclerView.Adapter {
    private List<IVerse> mVerses = new ArrayList<>();
    private String copyrightText;

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == VerseViewHolder.getType()) {
            return new VerseViewHolder(parent);
        } else if (viewType == CopyrightViewHolder.getType()) {
            return new CopyrightViewHolder(parent);
        }

        throw new IllegalArgumentException("No viewholder that matches the requested type");
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof VerseViewHolder) {
            ((VerseViewHolder) holder).setVerseText(mVerses.get(position).getText(holder.itemView.getContext()));
        } else if (holder instanceof CopyrightViewHolder) {
            ((CopyrightViewHolder) holder).setText(copyrightText);
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (position < mVerses.size()) {
            return VerseViewHolder.getType();
        } else if (position == mVerses.size()) {
            return CopyrightViewHolder.getType();
        }

        return -1;
    }

    @Override
    public int getItemCount() {
        return mVerses.size() + (copyrightText == null ? 0 : 1); //account for the copyright item
    }

    public void updateVerses(List<IVerse> verses) {
        mVerses.clear();
        mVerses.addAll(verses);

        if (!verses.isEmpty()) {
            copyrightText = verses.get(0).getCopyright();
        }

        notifyDataSetChanged();
    }
}

File: app/src/main/kotlin/com/pnuema/bible/android/ui/adapters/VersesAdapter.kt
package com.pnuema.bible.android.ui.adapters

import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil.ItemCallback
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.pnuema.bible.android.data.IVerse
import com.pnuema.bible.android.ui.viewholders.CopyrightViewHolder
import com.pnuema.bible.android.ui.viewholders.VerseViewHolder
import com.pnuema.bible.android.ui.viewstates.VerseViewState

class VersesAdapter : ListAdapter<VerseViewState, RecyclerView.ViewHolder>(
    object : ItemCallback<VerseViewState>() {
        override fun areItemsTheSame(oldItem: VerseViewState, newItem: VerseViewState): Boolean = false
        override fun areContentsTheSame(oldItem: VerseViewState, newItem: VerseViewState): Boolean = false
    }
) {
    private lateinit var copyrightText: String

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            VerseViewHolder.type -> VerseViewHolder(parent)
            CopyrightViewHolder.type -> CopyrightViewHolder(parent)
            else -> throw IllegalArgumentException("No view holder that matches the requested type")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder) {
            is VerseViewHolder -> holder.bind(currentList[position])
            is CopyrightViewHolder -> holder.bind(copyrightText)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return when {
            position < currentList.size -> VerseViewHolder.type
            position == currentList.size -> CopyrightViewHolder.type
            else -> -1
        }
    }

    override fun getItemCount(): Int {
        return currentList.size + if (!::copyrightText.isInitialized) 0 else 1 //account for the copyright item
    }
}
